### PR TITLE
Feature: Add GitHub groups (orgs/teams) support 

### PR DESCRIPTION
**Pull request by tuunit**
## Motivation and Context
GitHub's hierarchy is structured by organization and teams. As of now, in version 7.4.0 this hierarchy can only be used for restricting access but the details about a users membership are not passed through to the underlying applications behind the oauth2-proxy. Therefore this PRs aim is to extend the GitHub provider with the functionality to expose the users membership. As the session storage already provides the functionality to store "groups" and forward them as headers, this is a valid feature / extension and in line with other providers.

## Description of the implementation details
I refactored the GitHub provider quite extensively. To ensure backwards compatibility all the restrictions and checks are done as before. The only logical change is that instead of just fetching the org and team information when an organisation or team restriction is set, the membership information is always fetched from GitHub and added to the sessions group storage.

## How Has This Been Tested?
All existing GitHub tests have been corrected to accommodate the new code structure. Manual testing has been done and I already use this version of the implementation in a production environment.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [X] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by tuunit** [comment]
@JoelSpeed new PR for #1928

@hoax I would highly appreciate a code review from your side and especially check that I didn't miss anything when rebasing our commits.

**Comment by hoax** [comment]
@tuunit, looks good, but did not test it yet

**Review by tuunit — COMMENTED** [review]
Reminder to myself: Rebase and fix merge conflicts

**Comment by tuunit** [comment]
@kvanzuijlen can you give this a review? :)

File: providers/github.go [hunk 1]
```diff
@@ -31,7 +32,8 @@ var _ Provider = (*GitHubProvider)(nil)
 
 const (
 	githubProviderName = "GitHub"
-	githubDefaultScope = "user:email"
+	githubDefaultScope = "user:email read:org"
```
Comment (kvanzuijlen) [inline providers/github.go:35]: Any reason why this is now the default? Wouldn't this be a breaking change?

Comment (tuunit) [reply providers/github.go:35]: As it is adding and not removing a scope to the default it is not a breaking change. It was already necessary to add `read:org` when you wanted to get information about the teams and orgs of a user.

**Review by kvanzuijlen — APPROVED** [review]
LGTM

**Review by JoelSpeed — COMMENTED** [review]
Is this implemented in a backwards compatible way?

**Review by JoelSpeed — APPROVED** [review]

File: providers/github.go (hunk 1 above)
Comment (etho201) [reply providers/github.go:35]: It's a breaking change if/when a user specified the scope. For example, I was using this flag `--scope=user:email` and this broke things after updating to `7.6.0`. To fix this, I had to either remove my manual setting of the flag or add the new setting `user:email read:org`.

- https://github.com/etho201/docker-pi-stacks/commit/b16eb1b87e83324b745b7cba2a713ba347513e86
- https://github.com/etho201/docker-pi-stacks/commit/22b1d0f18e787fc9a09997089a123091ce550600

Comment (tuunit) [reply providers/github.go:35]: @etho201 I'm sorry to hear that you faced this issue. I thought we specifically tested this scenario. Can you share the error you encountered and your full configuration?

**Comment by jacek-jablonski** [comment]
> Is this implemented in a backwards compatible way?

Nope, it blew away our production.

**Comment by tuunit** [comment]
> > Is this implemented in a backwards compatible way?
> 
> Nope, it blew away our production.

I've tested it with 3 different scenarios and another user tested the feature branch as well so I'm a bit lost what happened on your end.

Can you please share your config and which errors you encountered?

**Comment by jacek-jablonski** [comment]
Hi,
config by args:
```
--provider=github --scope="read:user" --email-domain=* --github-team="xxx/devs" --upstream=http://argo-server:2746 --http-address=0.0.0.0:4180 --reverse-proxy
```
`OAUTH2_PROXY_CLIENT_ID` and `OAUTH2_PROXY_CLIENT_SECRET` set by env vars.

Error was:
```
[2024/02/21 14:18:43] [oauthproxy.go:888] Error creating session during OAuth2 callback: unexpected status "403": {"message":"You need at least read:org scope or user scope to list your organizations.","documentation_url":"https://docs.github.com/rest/orgs/orgs#list-organizations-for-the-authenticated-user"}
```

**Comment by etho201** [comment]
> Hi, config by args:
> 
> ```
> --provider=github --scope="read:user" --email-domain=* --github-team="xxx/devs" --upstream=http://argo-server:2746 --http-address=0.0.0.0:4180 --reverse-proxy
> ```
> 
> `OAUTH2_PROXY_CLIENT_ID` and `OAUTH2_PROXY_CLIENT_SECRET` set by env vars.
> 
> Error was:
> 
> ```
> [2024/02/21 14:18:43] [oauthproxy.go:888] Error creating session during OAuth2 callback: unexpected status "403": {"message":"You need at least read:org scope or user scope to list your organizations.","documentation_url":"https://docs.github.com/rest/orgs/orgs#list-organizations-for-the-authenticated-user"}
> ```

I had the same problem. It's a breaking change if/when a user specified the scope. For example, I was using this flag `--scope=user:email` and this broke things after updating to 7.6.0. To fix this, I had to either remove my manual setting of the flag or add the new setting `user:email read:org`. If you don't specify, the default value is now `user:email read:org`. See here for more: https://github.com/oauth2-proxy/oauth2-proxy/pull/2196/files#r1492008964

**Comment by tuunit** [comment]
@etho201 thanks for sharing. Indeed that is the issue we will update it in the CHANGELOG and I'll talk to @JoelSpeed if we can get a bugfix released.

**Comment by tuunit** [comment]
@etho201 @jacek-jablonski 
I'm sorry for the inconvenience :/

**Comment by jacek-jablonski** [comment]
> @etho201 @jacek-jablonski 
> 
> I'm sorry for the inconvenience :/

No problem at all. Thanks for the great work 🙂

**Pull request merged**